test/utils/wasm3_engine.cpp

```diff
@@ -0,0 +1,100 @@
+#include "source/wasm3.h"
+#include <test/utils/wasm_engine.hpp>
+#include <cassert>
+#include <cstring>
+
+namespace fizzy::test
+{
+class Wasm3Engine : public WasmEngine
+{
+    IM3Environment m_env{nullptr};
+    IM3Runtime m_runtime{nullptr};
+
+public:
+    Wasm3Engine() : m_env(m3_NewEnvironment()) {}
+    ~Wasm3Engine()
+    {
+        if (m_runtime)
+            m3_FreeRuntime(m_runtime);
+        m3_FreeEnvironment(m_env);
+    }
+
+    void parse(bytes_view input) final;
+    std::optional<FuncRef> find_function(std::string_view name) const final;
+    void instantiate() final;
+    fizzy::bytes_view get_memory() const final;
+    void set_memory(fizzy::bytes_view memory) final;
+    Result execute(FuncRef func_ref, const std::vector<uint64_t>& args) final;
+};
+
+std::unique_ptr<WasmEngine> create_wasm3_engine()
+{
+    return std::make_unique<Wasm3Engine>();
+}
+
+void Wasm3Engine::parse(bytes_view input)
+{
+    // Replace runtime (e.g. instance + module)
+    if (m_runtime != nullptr)
+        m3_FreeRuntime(m_runtime);
+
+    // The 64k stack size comes from wasm3/platorms/app
+    m_runtime = m3_NewRuntime(m_env, 64 * 1024, NULL);
```

Comment:
Could move Runtime+LoadModule to instantiate.

Reply:
The `m3_ParseModule()` looks like parsing only. It might be good to only leave this one here.

On the other hand it seems impossible to split parsing and instantiation in wabt, but we might need to merge `parse()` and `instantiate()`.

Reply:
Btw I tried this split, but it was crashing. Didn't bother figuring out why it crashed.